File: backend/api/databases/databases.py
"""Database HTTP controllers.

Thin layer: validates input, delegates to the service and returns the standard
envelope. Domain errors raised by the service are translated by the global
error handler in ``middlewares/errors.py``.
"""
import json

from flask import Blueprint, request
from flask_jwt_extended import jwt_required
from marshmallow import ValidationError

from middlewares.auth import get_current_user
from middlewares.responses import error, success
from schemas.database_schema import (
    DatabaseFilterSchema,
    DatabaseSortSchema,
)
from services.database_service import (
    add_item,
    add_property,
    create_database,
    delete_database,
    delete_item,
    delete_property,
    get_database,
    list_databases,
    update_database,
    update_item,
    update_property,
)

databases_bp = Blueprint("databases", __name__, url_prefix="/databases")


def _parse_query_list(name: str) -> list | None:
    raw = request.args.get(name)
    if not raw:
        return None
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        raise ValidationError(f"{name} must be valid JSON")


def _validate_filters_sorts() -> tuple[list | None, list | None]:
    filters = _parse_query_list("filter")
    sorts = _parse_query_list("sort")
    if filters is not None:
        filters = DatabaseFilterSchema(many=True).load(filters)
    if sorts is not None:
        sorts = DatabaseSortSchema(many=True).load(sorts)
    return filters, sorts


@databases_bp.post("")
@jwt_required()
def create():
    user = get_current_user()
    payload = request.get_json(force=True, silent=True) or {}
    database = create_database(user.id, payload)
    return success("Database created", database, 201)


@databases_bp.get("/workspace/<int:workspace_id>")
@jwt_required()
def list_for_workspace(workspace_id: int):
    user = get_current_user()
    databases = list_databases(user.id, workspace_id)
    return success("Databases retrieved", databases)


@databases_bp.get("/<int:database_id>")
@jwt_required()
def get_one(database_id: int):
    user = get_current_user()
    filters, sorts = _validate_filters_sorts()
    database = get_database(user.id, database_id, filters, sorts)
    return success("Database retrieved", database)


@databases_bp.put("/<int:database_id>")
@jwt_required()
def update(database_id: int):
    user = get_current_user()
    payload = request.get_json(force=True, silent=True) or {}
    database = update_database(user.id, database_id, payload)
    return success("Database updated", database)


@databases_bp.delete("/<int:database_id>")
@jwt_required()
def delete(database_id: int):
    user = get_current_user()
    delete_database(user.id, database_id)
    return success("Database deleted", None)


@databases_bp.post("/<int:database_id>/properties")
@jwt_required()
def create_property(database_id: int):
    user = get_current_user()
    payload = request.get_json(force=True, silent=True) or {}
    prop = add_property(user.id, database_id, payload)
    return success("Property added", prop, 201)


@databases_bp.put("/<int:database_id>/properties/<int:prop_id>")
@jwt_required()
def edit_property(database_id: int, prop_id: int):
    user = get_current_user()
    payload = request.get_json(force=True, silent=True) or {}
    prop = update_property(user.id, database_id, prop_id, payload)
    return success("Property updated", prop)


@databases_bp.delete("/<int:database_id>/properties/<int:prop_id>")
@jwt_required()
def remove_property(database_id: int, prop_id: int):
    user = get_current_user()
    delete_property(user.id, database_id, prop_id)
    return success("Property deleted", None)


@databases_bp.post("/<int:database_id>/items")
@jwt_required()
def create_item(database_id: int):
    user = get_current_user()
    payload = request.get_json(force=True, silent=True) or {}
    item = add_item(user.id, database_id, payload)
    return success("Item added", item, 201)


@databases_bp.put("/<int:database_id>/items/<int:item_id>")
@jwt_required()
def edit_item(database_id: int, item_id: int):
    user = get_current_user()
    payload = request.get_json(force=True, silent=True) or {}
    item = update_item(user.id, database_id, item_id, payload)
    return success("Item updated", item)


@databases_bp.delete("/<int:database_id>/items/<int:item_id>")
@jwt_required()
def remove_item(database_id: int, item_id: int):
    user = get_current_user()
    delete_item(user.id, database_id, item_id)
    return success("Item deleted", None)

File: frontend/src/services/databases.ts
import { ApiResponse, Database, DatabaseItem, DatabaseProperty } from "@/types";
import api from "@/lib/axios";

export interface CreateDatabasePayload {
  workspace_id: number;
  name: string;
  icon?: string;
  properties?: Array<{
    name: string;
    type: string;
    options?: { choices?: string[] };
    position?: number;
  }>;
}

export const databaseService = {
  async list(workspaceId: number): Promise<Database[]> {
    const { data } = await api.get<ApiResponse<Database[]>>(
      `/databases/workspace/${workspaceId}`
    );
    return (data.data as Database[]) || [];
  },

  async get(databaseId: number): Promise<Database> {
    const { data } = await api.get<ApiResponse<Database>>(`/databases/${databaseId}`);
    return data.data as Database;
  },

  async create(payload: CreateDatabasePayload): Promise<Database> {
    const { data } = await api.post<ApiResponse<Database>>("/databases", payload);
    return data.data as Database;
  },

  async update(
    databaseId: number,
    payload: { name?: string; icon?: string }
  ): Promise<Database> {
    const { data } = await api.put<ApiResponse<Database>>(
      `/databases/${databaseId}`,
      payload
    );
    return data.data as Database;
  },

  async remove(databaseId: number): Promise<void> {
    await api.delete(`/databases/${databaseId}`);
  },

  async addProperty(
    databaseId: number,
    payload: { name: string; type: string; options?: { choices?: string[] }; position?: number }
  ): Promise<DatabaseProperty> {
    const { data } = await api.post<ApiResponse<DatabaseProperty>>(
      `/databases/${databaseId}/properties`,
      payload
    );
    return data.data as DatabaseProperty;
  },

  async updateProperty(
    databaseId: number,
    propId: number,
    payload: { name?: string; type?: string; options?: { choices?: string[] }; position?: number }
  ): Promise<DatabaseProperty> {
    const { data } = await api.put<ApiResponse<DatabaseProperty>>(
      `/databases/${databaseId}/properties/${propId}`,
      payload
    );
    return data.data as DatabaseProperty;
  },

  async removeProperty(databaseId: number, propId: number): Promise<void> {
    await api.delete(`/databases/${databaseId}/properties/${propId}`);
  },

  async addItem(
    databaseId: number,
    values: Array<{ property_id: number; value: string | number | null }>,
    position?: number
  ): Promise<DatabaseItem> {
    const { data } = await api.post<ApiResponse<DatabaseItem>>(
      `/databases/${databaseId}/items`,
      { values, position }
    );
    return data.data as DatabaseItem;
  },

  async updateItem(
    databaseId: number,
    itemId: number,
    values?: Array<{ property_id: number; value: string | number | null }>,
    position?: number
  ): Promise<DatabaseItem> {
    const { data } = await api.put<ApiResponse<DatabaseItem>>(
      `/databases/${databaseId}/items/${itemId}`,
      { values, position }
    );
    return data.data as DatabaseItem;
  },

  async removeItem(databaseId: number, itemId: number): Promise<void> {
    await api.delete(`/databases/${databaseId}/items/${itemId}`);
  },
};

export const TASKS_PRESET: CreateDatabasePayload = {
  workspace_id: 0,
  name: "Tarefas",
  icon: "✅",
  properties: [
    { name: "Título", type: "text", position: 0 },
    {
      name: "Status",
      type: "status",
      position: 1,
      options: { choices: ["A fazer", "Em andamento", "Concluído"] },
    },
    { name: "Prazo", type: "date", position: 2 },
    { name: "Responsável", type: "text", position: 3 },
  ],
};
